traitsui/qt4/datetime_editor.py

```diff
@@ -15,60 +15,79 @@
 """
 
 
-import datetime
-
-from pyface.qt import QtGui
+from pyface.qt import QtGui, qt_api
 from pyface.qt.QtCore import QDateTime
-from traits.api import Instance
+from traits.api import Datetime, on_trait_change
 
 from .editor import Editor
 from .editor_factory import ReadonlyEditor as BaseReadonlyEditor
 
 
+
 class SimpleEditor(Editor):
     """ Simple Traits UI time editor that wraps QDateTimeEdit.
     """
 
     #: the earliest datetime allowed by the editor
-    minimum_datetime = Instance(datetime.datetime)
+    minimum_datetime = Datetime
 
     #: the latest datetime allowed by the editor
-    maximum_datetime = Instance(datetime.datetime)
+    maximum_datetime = Datetime
 
     def init(self, parent):
         """ Finishes initializing the editor by creating the underlying toolkit
             widget.
         """
-        self.control = QtGui.QDateTimeEdit()
+        # set min and max early, don't wait for editor sync
         self.minimum_datetime = self.factory.minimum_datetime
         self.maximum_datetime = self.factory.maximum_datetime
-        self.update_date_range()
+
+        self.control = QtGui.QDateTimeEdit()
         self.control.dateTimeChanged.connect(self.update_object)
+        self.update_date_range()
 
     def update_editor(self):
         """ Updates the editor when the object trait changes externally to the
             editor.
         """
         value = self.value
-        if value:
-            q_datetime = QDateTime(value)
-            self.control.setTime(q_datetime)
+        if self.value:
+            if self.minimum_datetime and self.minimum_datetime > value:
+                value = self.minimum_datetime
+            elif self.maximum_datetime and self.value > self.maximum_datetime:
+                value = self.maximum_datetime
```

Comment:
This sets the view to the minimum/maximum value without updating the trait value being edited.
Looks like we need to assign the value to `self.value`

Reply:
It loops back: `value` gets set as the value in the editor, which in turn fires a signal, which sets the trait value, but in a clear way.  If we set `self.value` I think we would end up calling this method twice.

Reply:
Hmmm maybe it is the backend I was using (I was testing with PyQt5 but I have not tried PySide2 nor PyQt5), the signal is not set and the following tests are failing until I assign `self.value` here:
```

    def test_datetime_editor_simple_with_minimum_datetime_out_of_bound(self):
        # Test when out-of-bound value is given, it is reset to the bounded
        # value.
        minimum_datetime = datetime.datetime(2000, 1, 1)
        editor_factory = DatetimeEditor(
            minimum_datetime=minimum_datetime,
        )
        view = get_date_time_simple_view(editor_factory)
        instance = InstanceWithDatetime()
        with self.launch_editor(instance, view) as editor:
            instance.date_time = datetime.datetime(1980, 1, 1)
            displayed_value = to_datetime(editor.control.dateTime())

        self.assertEqual(displayed_value, minimum_datetime)
        self.assertEqual(instance.date_time, minimum_datetime)


    def test_datetime_editor_simple_with_maximum_datetime_out_of_bound(self):
        maximum_datetime = datetime.datetime(2000, 1, 1)
        editor_factory = DatetimeEditor(
            maximum_datetime=maximum_datetime,
        )
        view = get_date_time_simple_view(editor_factory)
        instance = InstanceWithDatetime()
        with self.launch_editor(instance, view) as editor:
            # out-of-bound
            instance.date_time = datetime.datetime(2020, 1, 1)
            displayed_value = to_datetime(editor.control.dateTime())

        self.assertEqual(displayed_value, maximum_datetime)
        self.assertEqual(instance.date_time, maximum_datetime)
```

Reply:
It would be helpful if I also present the failures here, there they are:
```
======================================================================
FAIL: test_datetime_editor_simple_with_maximum_datetime_out_of_bound (traitsui.tests.editors.test_datetime_editor.TestDatetimeEditorQt)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/traitsui/traitsui/tests/editors/test_datetime_editor.py", line 100, in test_datetime_editor_simple_with_maximum_datetime_out_of_bound
    self.assertEqual(instance.date_time, maximum_datetime)
AssertionError: datetime.datetime(2020, 1, 1, 0, 0) != datetime.datetime(2000, 1, 1, 0, 0)

======================================================================
FAIL: test_datetime_editor_simple_with_minimum_datetime_out_of_bound (traitsui.tests.editors.test_datetime_editor.TestDatetimeEditorQt)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/traitsui/traitsui/tests/editors/test_datetime_editor.py", line 71, in test_datetime_editor_simple_with_minimum_datetime_out_of_bound
    self.assertEqual(instance.date_time, minimum_datetime)
AssertionError: datetime.datetime(1980, 1, 1, 0, 0) != datetime.datetime(2000, 1, 1, 0, 0)
```